### PR TITLE
PLT-235 Implement outgoing webhooks.

### PR DESCRIPTION
**Pull request by jwilander**
In this PR:

* Added new OutgoingWebhook model with appropriate functions
* Added new OutgoingWebhooks table to store with appropriate functions
* Added API services dealing with outgoing webhooks
* Updated store index creation to use predefined types
* Replaced the rarely used doUpdateLastViewed parameter in api.CreatePost() with a useful triggerWebhook parameter
* Added a fireAndForgetTriggerWebhooks function to submit post events to relevant webhooks
* Lots of related unit tests

The implementation is slightly different than discussed. The original plan was to have an `Event` column in the database and select based off that. Turns out multiple event types are beyond the scope of the currently required functionality, and all we need is new post events. In the future if we need to support more events we can add functionality for that without too much trouble.

### CODE REVIEW

File: api/post.go [hunk 1]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
...
 	return rpost, nil
 }
 
+func fireAndForgetWebhookEvent(c *Context, post *model.Post) {
+
+	go func() {
+
+		chchan := Srv.Store.Webhook().GetOutgoingByChannel(post.ChannelId)
+
+		firstWord := strings.Split(post.Message, " ")[0]
```
Comment (coreyhulen) [inline api/post.go]: can this fail?  img posts with no messages.

File: api/post.go [hunk 2]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
 	var rpost *model.Post
 	if result := <-Srv.Store.Post().Save(post); result.Err != nil {
 		return nil, result.Err
-	} else if doUpdateLastViewed && (<-Srv.Store.Channel().UpdateLastViewedAt(post.ChannelId, c.Session.UserId)).Err != nil {
-		return nil, result.Err
 	} else {
 		rpost = result.Data.(*model.Post)
 
 		fireAndForgetNotifications(rpost, c.Session.TeamId, c.GetSiteURL())
```
Comment (coreyhulen) [inline api/post.go]: why isn't email notifications written as a hook?

File: api/post.go [hunk 3]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
...
+
+		if result := <-thchan; result.Err != nil {
+			l4g.Error("Encountered error getting webhook by trigger word, err=%v", result.Err)
+			return
+		} else {
+			hooks = append(hooks, result.Data.([]*model.OutgoingWebhook)...)
+		}
+
+		cchan := Srv.Store.Channel().Get(post.ChannelId)
+		uchan := Srv.Store.User().Get(post.UserId)
```
Comment (coreyhulen) [inline api/post.go]: is user and channel already loaded calling function?  maybe just pass this in.

File: api/post.go [hunk 4]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
...
+
+		tchan := make(chan *model.Team)
+		isTeamBeingFetched := make(map[string]bool)
+		for _, hook := range hooks {
+			if !isTeamBeingFetched[hook.TeamId] {
+				isTeamBeingFetched[hook.TeamId] = true
+				go func() {
+					if result := <-Srv.Store.Team().Get(hook.TeamId); result.Err != nil {
```
Comment (coreyhulen) [inline api/post.go]: why is this needed?  I can only have 1 team correct?  user.TeamId or session.TeamId

File: api/post.go [hunk 5]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
...
+		for range isTeamBeingFetched {
+			team := <-tchan
+			if team != nil {
+				teams[team.Id] = team
+			}
+		}
+
+		var channel *model.Channel
+		if result := <-cchan; result.Err != nil {
+			l4g.Error("Encountered error getting channel, channel_id=%s, err=%v", post.ChannelId, result.Err)
```
Comment (coreyhulen) [inline api/post.go]: no return on error?  what happens later?

File: api/post.go [hunk 6]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
...
+		} else {
+			channel = result.Data.(*model.Channel)
```
Comment (coreyhulen) [inline api/post.go]: no return on error?  what happens later?

File: api/post.go [hunk 7]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
...
+		}
+
+		var user *model.User
+		if result := <-uchan; result.Err != nil {
+			l4g.Error("Encountered error getting user, user_id=%s, err=%v", post.UserId, result.Err)
+		} else {
+			user = result.Data.(*model.User)
```
Comment (coreyhulen) [inline api/post.go]: no return on error?  what happens later?

File: api/post.go (hunk 4 above)
Comment (jwilander) [inline api/post.go]: Ah, yeah. Not sure why I thought there would be multiple teams. Cleaning up..

File: api/post.go (hunk 5 above)
Comment (jwilander) [inline api/post.go]: This is on purpose since we can still post the message without the channel (we just pull it for the name). This is rarely going to fail since the channel must exist as it's confirmed in createPost. 0/5 if you a return here anyways

File: store/sql_store.go [hunk 8]
```diff
@@ -371,8 +381,10 @@ func (ss SqlStore) createIndexIfNotExists(indexName string, tableName string, co
 		}
 
 		query := ""
-		if fullText {
+		if indexType == INDEX_TYPE_FULL_TEXT {
 			query = "CREATE INDEX " + indexName + " ON " + tableName + " USING gin(to_tsvector('english', " + columnName + "))"
+		} else if indexType == INDEX_TYPE_PATTERN {
+			query = "CREATE INDEX " + indexName + " ON " + tableName + " (" + columnName + " text_pattern_ops)"
```
Comment (coreyhulen) [inline store/sql_store.go]: This is for postgres?  what about mysql?

File: api/post.go (hunk 7 above)
Comment (jwilander) [inline api/post.go]: Similar to above comment, the POST can still be made without the user object it just will be missing the username

File: api/post.go (hunk 1 above)
Comment (jwilander) [inline api/post.go]: Split will return an array of size 1 with a blank string at the minimum, so no it won't fail.

File: store/sql_store.go (hunk 8 above)
Comment (coreyhulen) [inline store/sql_store.go]: why go through the hassle of using the db when it's such a small number of webhooks ??.  Just return the all and post filter in the server.  Since the first word is constantly changing you never get any db query cache.  You could just hold the hooks in memory.

Comment (jwilander) [inline store/sql_store.go]: MySQL we use full text search for pattern matching

File: api/post.go [hunk 9]
```diff
@@ -128,18 +132,142 @@ func CreatePost(c *Context, post *model.Post, doUpdateLastViewed bool) (*model.P
...
+				p.Set("channel_id", post.ChannelId)
+				if channel != nil {
+					p.Set("channel_name", channel.Name)
+				}
+
+				p.Set("timestamp", strconv.FormatInt(post.CreateAt/1000, 10))
+
+				p.Set("user_id", post.UserId)
+				if user != nil {
+					p.Set("user_name", user.Username)
```
Comment (coreyhulen) [inline api/post.go]: this line will crash since user==nil ?

Comment (coreyhulen) [inline api/post.go]: nvm

File: api/post.go (hunk 3 above)
Comment (jwilander) [inline api/post.go]: No, unfortunately they're not pulled in the parent function

File: api/post.go (hunk 2 above)
Comment (coreyhulen) [inline api/post.go]: email notifications are loading 

uchan := Srv.Store.User().GetProfiles(teamId)
		echan := Srv.Store.Channel().GetMembers(post.ChannelId)
		cchan := Srv.Store.Channel().Get(post.ChannelId)
		tchan := Srv.Store.Team().Get(teamId)

Seems like we're loading stuff twice for no reason.  Probably could use a refactor.

Comment (jwilander) [inline api/post.go]: Yeah, agreed. Is this ok to do in another PR though?

**Comment by jwilander** [comment]
Closing to make changes, will resubmit
